Allow JTAG scan information to be encoded in targets

seanmlyons22 opened the pull request:
The scan chain information structure is a direct implementation of [Scan Chain](https://open-cmsis-pack.github.io/Open-CMSIS-Pack-Spec/main/html/sdf_pg.html#sdf_element_scanchain) from CMSIS pack's System Description File (SDF)

It aims to resolve #1682 by adding scan chain information to the target, passing scan chain information to the probe, and finally allowing the `CmsisDap` probe to use scan chain information if its provided.

Having scan chain information opens the possibly for the following (these are open ideas, not yet implemented)

 * Refactoring of existing `stlink` and other probes that are already working with `ir_len`. Clean them up, use scan chain instead.
 * Updating `target-gen` to parse scan chain information from the SDF if its included in a CMSIS pack

Places where I am looking for feedback:

1. [ScanChain definition](https://github.com/seanmlyons22/probe-rs/blob/specify_ir_lengths/probe-rs-target/src/chip.rs#L6-L39). 
1. Should I have so many optional fields? Its defined that way in CMSIS pack, but makes the code a bit mess. Similiar is the `Device` vs. `Dap` distinction even necessary?
1. How about the connection between Target and Probe? Now I am copying the scan chain into probe, but I am not really satisfied with how this came out...it feels clunky. End of the day, the interface (e.g. `CmsisDap`) just needs to know the scan chain for its currently connected target.

Yatekii commented on probe-rs/src/probe.rs:
```diff
@@ -538,6 +548,19 @@ pub trait DebugProbe: Send + fmt::Debug {
+    /// This allows the probe to know which TAPs are in the scan chain and their
+    /// position and IR lengths.
+    ///
+    /// If the scan chain is provided, and the selected protocol is JTAG, the
```
This might be misleading because depending on when one calls `set_scan_chain`, the call might not have any effect because the change will not be applied. Maybe we should clarify that bit?

Yatekii commented on probe-rs/src/probe.rs:
```diff
@@ -419,6 +420,15 @@ impl Probe {
         }
     }
 
+    /// Configure the scan chain to use.
```
Can we add a refrence to the trait method here so people get the full picture or even maybe copy paste the docs and add the bit that this is only callable before attaching :)

seanmlyons22 replied:
Definitely. Good ideas.

Yatekii commented on probe-rs/src/config/target.rs:
```diff
@@ -40,6 +42,8 @@ pub struct Target {
     pub(crate) source: TargetDescriptionSource,
     /// Debug sequences for the given target.
     pub debug_sequence: DebugSequence,
+    /// Description of scan chain
```
```suggestion
    /// Description of the scan chain
```

Yatekii commented on probe-rs-target/src/chip.rs:
```diff
@@ -1,6 +1,42 @@
+    /// List of protocols via which the DAP can be use. Possible values are SWD, JTAG, and cJTAG.
+    pub protocols: Option<Vec<String>>,
+    /// Specifies the SW-DPv2 TARGETSEL register value that selects this DAP in a serial-wire debug multi-drop system
+    pub targetsel: Option<u8>,
```
We already have a field for this. Can we not reuse that one? Is there a way to unify the two? I think the only chip currently really using it is the rp2040 :)

seanmlyons22 replied:
I tried to define the struct as below, but then I realized that `probe-rs-target` doesn't have dependencies on `probe-rs` only the other way around. Since `DpAddress`, `PortType`, and `WireProtocol` are defined in probe-rs, I didn't add that dependency to avoid a loop.

Instead of doing that, I just paired down the definition of `ScanChainElement`


```rust
pub struct ScanChainElement {
    /// Unique name of the DAP
    pub name: Option<String>,
    /// Specifies the type of the DAP (e.g. ARMCS-DP)
    pub element_type: Option<PortType>,
    /// Specifies the IR length of the DAP (default value: 4).
    pub ir_len: Option<u8>,
    /// List of protocols via which the DAP can be use. Possible values are SWD, JTAG, and cJTAG.
    pub protocols: Option<Vec<WireProtcol>>,
    /// Specifies the SW-DPv2 TARGETSEL register value that selects this DAP in a serial-wire debug multi-drop system
    pub targetsel: Option<DpAddress>,
}
```

Yatekii replied:
I think we would have to define the types in `probe-rs-target` and have the same type in `probe-rs` with a simple `From` conversion impl.
I think proper typing should be applied :) What do you think? :)

Yatekii replied:
Sounds good to me!

Yatekii commented on probe-rs-target/src/chip.rs:
```diff
@@ -1,6 +1,42 @@
+    /// Unique name of the DAP
+    pub name: Option<String>,
+    /// Specifies the type of the DAP (e.g. ARMCS-DP)
+    pub dap_type: Option<String>,
```
I am a bit confused by the type of the DAP. Usually there is 1-N DAPs that have 1-M[n] DPs each. Which then have 1-O[m] APs.
At leats that's what I thought.
It would be great if this was an enum, even if it might not be conclusive at the start.

seanmlyons22 replied:
We discussed it a bit in the chat, but the idea I originally had was to distinguish between JTAG chain elements that were debuggable or not. For example in the TI chips that I am working with there is a TAP for the ICEPICK router that is not debuggable, so probe-rs cannot connect to it. I thought to use `device` there. This was also made to match 1:1 with what CMSIS did here:

[scan_chain in CMSIS SDF file](https://open-cmsis-pack.github.io/Open-CMSIS-Pack-Spec/main/html/sdf_pg.html#sdf_element_scanchain). They have DAP and device.

Eventually I thought to keep the fields 1:1 between probe-rs and the SDF file for easy parsing, but I think that makes the code a bit messier and adds unnecessary complexity, probe-rs doesn't really care about any nondebuggable targets outside of their role in the scan chain, so we don't need this enum.

Yatekii replied:
Fair enough, that sounds good to me! If we ever need more functionality later on, we can upgrade it. I think it would be interesting to use the ICEPICK router through probe-rs, but it's a bit out of scope for now :)

Yatekii commented on probe-rs-target/src/chip.rs:
```diff
@@ -1,6 +1,42 @@
+    /// Specifies the IR length of the DAP (default value: 4).
+    pub ir_len: Option<u8>,
+    /// List of protocols via which the DAP can be use. Possible values are SWD, JTAG, and cJTAG.
+    pub protocols: Option<Vec<String>>,
```
Here we should use the `Protocol` enum :)

Yatekii commented on probe-rs-target/src/chip.rs:
```diff
@@ -1,6 +1,42 @@
+    /// Unique name of the device
+    pub name: Option<String>,
+    /// Specifies the type of the device (e.g. ARMCS-DP)
+    pub device_type: Option<String>,
```
Same comment as above :) What's the difference between a device and a DAP? Is the device the unification of multiple DAPs? then maybe we should have the structure as such? Again, I am not sure here, just raising some question to make your life harder ;P

Yatekii commented on probe-rs-target/src/chip.rs:
```diff
@@ -1,6 +1,42 @@
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub enum ScanChain {
+    /// Scan chain element is a DAP
+    Dap(ScanChainDap),
```
Maybe DAPs should actually be stored in a device and we don't need this enum?

seanmlyons22 replied:
You're right, that was just over complicating things. Removed.

Yatekii commented:
Thanks a lot! It's looking quite good I think - at least we have a clear path forward :) I raised a few questions :)
I am not entirely sure what the best structure is everywhere. probe-rs tries to map the physical representation as closely as possible, but that's not always the best. But it's definitely not great to keep the structure as in the XML - those formats are not optimal if you ask me.

1. I added a few comments :)
2. I think we should avoid options wherever possible, but we should also not take shortcuts and avoid it just because it's harder when it's actually necessary :) Also, I think we can make the DAPs nested under a device, but maybe we don't even need that and can simply go with the DAPs :) Maybe for scans of things other than DAPs, this distinction could be valuable in the future?
3. Is the scan chain something that's required during probe operation on the host side or only during setup? If it's only during setup, I would say we just pass it to the setup calls and don't store it? If it's unavoidable, I am fine storing it. It also makes it easier to get better debug logs if something goes wrong I guess.
Also, I guess for the whole chain, we might need multiple targets in the future, which would require us to change the session behavior. Maybe we worry about that lateron?

Thanks a lot for working on this!

seanmlyons22 commented:
Took a while to get back to this, but had another rework today. Summary:

1. Fixed all CI checks
1. Removed redundancy in `Dap` vs. `Device` enum values. Removed all unused fields in `ScanChainElement`. We don't follow exactly CMSIS-PACK SDF file, but that is fine. 
1. I didn't introduce a dependency on `probe-rs` from `probe-rs-target` so I skipped any struct members that would need definitions from there.

Things that I am still considering:

1. Try to ensure 0 copying or cloning on `scan_chain`.
2. Try to parse `scan_chain` from SDF file.
3. Have other probes leverage scan chain from target instead of just CMSIS DAP.

Things that IMO should be done in separate issues:

1. Cleanup all existing JTAG code (including some duplicate scan chain definitions) for example multi IR definitions.
2. Adding more values as needed to scan chain definition

Yatekii commented on probe-rs/src/probe.rs:
```diff
@@ -367,7 +368,7 @@ impl Probe {
     /// Selects the transport protocol to be used by the debug probe.
     pub fn select_protocol(&mut self, protocol: WireProtocol) -> Result<(), DebugProbeError> {
-        if !self.attached {
+        if self.attached {
```
this seems wrong. If the target is already attached, the protocol cannot be changed. What am I missing?

Yatekii commented on probe-rs/src/probe.rs:
```diff
@@ -420,6 +421,20 @@ impl Probe {
+        &mut self,
+        scan_chain: Vec<ScanChainElement>,
+    ) -> Result<(), DebugProbeError> {
+        if !self.attached {
```
In the description of this method in the Trait, you write that one has to be attched to call this method, yet, here you do the opposite check. Am I misunderstanding something?

seanmlyons22 replied:
No. you're absolutely right. That's embarrassing it looks like when cleaning up I did two things:

1. Changed the wrong function `select_protocol`. That shouldn't be changed in this PR at all.
1. Didn't change an existing bug in `set_scan_chain` which had the wrong logic as you pointed out.

Yatekii commented on probe-rs/src/probe/cmsisdap/mod.rs:
```diff
@@ -717,6 +720,14 @@ impl DebugProbe for CmsisDap {
+        if self.active_protocol() != Some(WireProtocol::Jtag) {
+            tracing::warn!("Setting Scan Chain is only supported in JTAG mode");
+        }
+        self.scan_chain = Some(scan_chain);
```
Maybe this method should also make a log that the scan chain is being set (or even better the location where it is actually configured), so when people fuck up the scanchain and probe-rs explodes weirdly later on, they get a hint :)

seanmlyons22 replied:
Definitely. Even better, I should leverage `DebugProbeError` here and propagate the error up so it doesn't just silently explode as you say!

Yatekii reviewed (commented):
Thanks a lot!

I added some thoughts here and there :)

Yatekii reviewed (approved):
Looking great, thank you!

cstrahan commented:
Nice work!

My one suggestion would be to call `jtag_scan` in both scenarios (whether IR lengths are/aren't specified), so the user can get a nice error when the scan chain is definitely messed up (maybe they forgot what they had connected to their probe, or if they have multiple devices daisy chained, then maybe they got the order of the devices mixed up, etc). Not absolutely critical, but I think it would be a nice to have (and this is why `jtag_scan` was designed to optionally take the expected IR lengths).
